File: ext/cpp_polygon_finder/PolygonFinder/src/polygon/reducers/RasterReducer.cpp
/*
 * RasterReducer.cpp
 *
 * Copyright (c) 2025-2026 Emanuele Cesaroni
 *
 * Licensed under the GNU Affero General Public License v3 (AGPLv3).
 * See the LICENSE file in this directory for the full license text.
 */

#include "RasterReducer.h"
#include <stdexcept>
#include <string>

RasterReducer::RasterReducer(std::vector<Point>& points, int versus)
  : Reducer(points),
    versus_(versus) {
}

void RasterReducer::reduce() {
  toRasterPolygon(points, versus_);
}

void RasterReducer::toRasterPolygon(std::vector<Point>& points, int versus) {
  const std::size_t n = points.size();

  if (n == 0) {
      return;
  }

  const bool outer = versus == Node::O;
  const int firstX = points.front().x;
  const int firstY = points.front().y;
  int previousX = points.back().x;
  int previousY = points.back().y;

  for (std::size_t i = 0; i < n; ++i) {
    const int currentX = points[i].x;
    const int currentY = points[i].y;
    int nextX;
    int nextY;
    if (i + 1 < n) {
      nextX = points[i + 1].x;
      nextY = points[i + 1].y;
    } else {
      nextX = firstX;
      nextY = firstY;
    }
    const Pole incomingSide = segmentPole(
      currentX - previousX,
      currentY - previousY,
      outer);
    const Pole outgoingSide = segmentPole(
      nextX - currentX,
      nextY - currentY,
      outer);
    int x = currentX;
    int y = currentY;
    const auto applySide = [&x, &y](Pole side) {
      switch (side) {
        case Pole::South:
          --y;
          break;

        case Pole::East:
          --x;
          break;

        case Pole::North:
        case Pole::West:
          break;
      }
    };
    applySide(incomingSide);
    if (outgoingSide != incomingSide) {
      applySide(outgoingSide);
    }
    points[i].x = x;
    points[i].y = y;
    previousX = currentX;
    previousY = currentY;
  }
}

RasterReducer::Pole RasterReducer::segmentPole(int dx, int dy, bool clockwise) {
  const int sx = sign(dx);
  const int sy = sign(dy);
  if (sx == 0 && sy == 1) {
    return clockwise ? Pole::East : Pole::West;
  }
  if (sx == 1 && sy == 0) {
    return clockwise ? Pole::North : Pole::South;
  }
  if (sx == 0 && sy == -1) {
    return clockwise ? Pole::West : Pole::East;
  }
  if (sx == -1 && sy == 0) {
    return clockwise ? Pole::South : Pole::North;
  }
  throw std::invalid_argument(
    "Non-axial segment dx=" +
    std::to_string(dx) +
    " dy=" +
    std::to_string(dy) +
    "!");
}

int RasterReducer::sign(int value) {
  return (value > 0) - (value < 0);
}

File: lib/contrek/reducers/raster_reducer.rb
# frozen_string_literal: true

module Contrek
  module Reducers
    class RasterReducer < Reducer
      def reduce!
        versus = @opts[:versus]
        to_raster_polygon!(@points, versus)
      end

      private

      def to_raster_polygon!(points, versus)
        n = points.size
        return points if n.zero?

        outer = versus == :o
        first_x = points[0][:x]
        first_y = points[0][:y]
        prev_x = points[-1][:x]
        prev_y = points[-1][:y]
        points.each_index do |i|
          curr_x = points[i][:x]
          curr_y = points[i][:y]
          if i + 1 < n
            next_x = points[i + 1][:x]
            next_y = points[i + 1][:y]
          else
            next_x = first_x
            next_y = first_y
          end
          in_side = segment_pole(
            curr_x - prev_x,
            curr_y - prev_y,
            outer
          )
          out_side = segment_pole(
            next_x - curr_x,
            next_y - curr_y,
            outer
          )
          x = curr_x
          y = curr_y
          [in_side, out_side].uniq.each do |side|
            case side
            when :south then y -= 1
            when :east then x -= 1
            end
          end
          points[i][:x] = x
          points[i][:y] = y
          prev_x = curr_x
          prev_y = curr_y
        end
        points
      end

      def segment_pole(dx, dy, clockwise)
        sx = dx <=> 0
        sy = dy <=> 0
        case [sx, sy]
        when [0, 1] then clockwise ? :east : :west
        when [1, 0] then clockwise ? :north : :south
        when [0, -1] then clockwise ? :west : :east
        when [-1, 0] then clockwise ? :south : :north
        else
          raise ArgumentError, "Non-axial segment dx=#{dx} dy=#{dy}!"
        end
      end
    end
  end
end
